File: llm-viz/tools/check_attention.py
"""Dump every self-attention intermediate from PyTorch for one sequence.

The end-to-end logit check already proves the whole forward pass agrees with
torch, but the visualization also *draws* the per-head internals (Q/K/V
slices, masked scores, softmax patterns, weighted V sums, the concat order).
This dump lets a companion JS check compare each of those tensors cell by
cell, so a bug that cancels out downstream would still be caught.

Run with:  python3 llm-viz/tools/check_attention.py
"""

import json
import math
import os

import numpy as np
import torch
import torch.nn.functional as F

from make_reference import load, PROMPT
from train_nano_gpt import BLOCK, C, N_HEAD, N_LAYER

HERE = os.path.dirname(os.path.abspath(__file__))
A = C // N_HEAD


def clean(mat):
    """NaN (the causal mask) is not valid JSON; use null for masked cells."""
    return [[None if math.isnan(v) else v for v in row] for row in mat]


def main():
    model = load()
    # same greedy decode the page performs
    seq = list(PROMPT)
    with torch.no_grad():
        while len(seq) < BLOCK:
            logits = model(torch.tensor([seq]))
            seq.append(int(logits[0, -1].argmax()))

        # walk the model manually with torch ops, capturing attention internals
        T = len(seq)
        pos = torch.arange(T)
        x = model.wte(torch.tensor(seq)) + model.wpe(pos)
        blocks = []
        for blk in model.blocks:
            n = blk.ln1(x)
            qkv = F.linear(n, blk.attn.qkv.weight, blk.attn.qkv.bias)
            q, k, v = qkv.split(C, dim=1)
            heads = []
            outs = []
            for h in range(N_HEAD):
                qh = q[:, h * A : (h + 1) * A]
                kh = k[:, h * A : (h + 1) * A]
                vh = v[:, h * A : (h + 1) * A]
                att = (qh @ kh.T) / math.sqrt(A)
                mask = torch.tril(torch.ones(T, T)).bool()
                att = att.masked_fill(~mask, float("nan"))
                sm = F.softmax(att.nan_to_num(float("-inf")), dim=-1)
                out = sm @ vh
                # the page draws masked cells as masked, not as probability 0
                sm = sm.masked_fill(~mask, float("nan"))
                heads.append(
                    {
                        "q": qh.tolist(),
                        "k": kh.tolist(),
                        "v": vh.tolist(),
                        "scores": clean(att.tolist()),  # masked above the diagonal
                        "pattern": clean(sm.tolist()),
                        "out": out.tolist(),
                    }
                )
                outs.append(out)
            combined = torch.cat(outs, dim=1)
            proj = F.linear(combined, blk.attn.proj.weight, blk.attn.proj.bias)
            resid = x + proj
            blocks.append(
                {
                    "heads": heads,
                    "combined": combined.tolist(),
                    "proj": proj.tolist(),
                    "resid": resid.tolist(),
                }
            )
            x = resid + blk.fcproj(F.gelu(blk.fc(blk.ln2(resid))))

    out_path = os.path.join(HERE, "attention_reference.json")
    with open(out_path, "w") as f:
        json.dump({"sequence": seq, "blocks": blocks}, f)
    print(f"wrote {out_path}: {N_LAYER} blocks x {N_HEAD} heads, T={len(seq)}")


if __name__ == "__main__":
    main()

File: llm-viz/tools/check_attention.js
/**
 * Compares every self-attention tensor the page draws against the PyTorch
 * dump from check_attention.py: per-head Q/K/V slices, masked scores, softmax
 * patterns and aggregates, weighted V sums, the head-major concat order, the
 * output projection and the residual.
 *
 * Run with:  python3 llm-viz/tools/check_attention.py && node llm-viz/tools/check_attention.js
 */
const fs = require("fs");
const path = require("path");

const assets = path.join(__dirname, "..", "assets");
global.window = global;
require(path.join(assets, "model.js"));

const meta = JSON.parse(fs.readFileSync(path.join(assets, "nano-gpt.json"), "utf8"));
const buf = fs.readFileSync(path.join(assets, "nano-gpt.bin"));
const all = new Float32Array(buf.buffer, buf.byteOffset, buf.byteLength / 4);
const w = {};
meta.tensors.forEach((t) => {
  const n = t.shape.reduce((a, b) => a * b, 1);
  w[t.name] = { data: all.subarray(t.offset, t.offset + n), shape: t.shape };
});
const model = { meta, cfg: meta.config, w };

const ref = JSON.parse(
  fs.readFileSync(path.join(__dirname, "attention_reference.json"), "utf8")
);
const run = LV.forward(model, ref.sequence);
const { C, headSize: A, nHeads: H, blockSize: T } = model.cfg;

let worst = { err: 0, where: "none" };
let failures = 0;

function compare(where, jsVal, refVal, tol = 2e-4) {
  const jsMasked = !isFinite(jsVal);
  const refMasked = refVal === null || !isFinite(refVal);
  if (jsMasked || refMasked) {
    if (jsMasked !== refMasked) {
      failures++;
      console.error(`MASK MISMATCH at ${where}: js=${jsVal} ref=${refVal}`);
    }
    return;
  }
  const err = Math.abs(jsVal - refVal);
  if (err > worst.err) worst = { err, where };
  if (err > tol) {
    failures++;
    if (failures < 10) console.error(`VALUE MISMATCH at ${where}: js=${jsVal} ref=${refVal}`);
  }
}

for (let b = 0; b < ref.blocks.length; b++) {
  const rb = ref.blocks[b];
  const jb = run.blocks[b];
  for (let h = 0; h < H; h++) {
    const rh = rb.heads[h];
    const jh = jb.heads[h];
    for (let t = 0; t < T; t++) {
      for (let a = 0; a < A; a++) {
        compare(`b${b}.h${h}.q[${t},${a}]`, jh.q.data[t * A + a], rh.q[t][a]);
        compare(`b${b}.h${h}.k[${t},${a}]`, jh.k.data[t * A + a], rh.k[t][a]);
        compare(`b${b}.h${h}.v[${t},${a}]`, jh.v.data[t * A + a], rh.v[t][a]);
        compare(`b${b}.h${h}.out[${t},${a}]`, jh.vOut.data[t * A + a], rh.out[t][a]);
        // the wide fused slab must agree with the per-head slices it is cut from
        compare(
          `b${b}.h${h}.qkvSliceQ[${t},${a}]`,
          jb.qkv.data[t * 3 * C + h * A + a],
          rh.q[t][a]
        );
        compare(
          `b${b}.h${h}.qkvSliceK[${t},${a}]`,
          jb.qkv.data[t * 3 * C + C + h * A + a],
          rh.k[t][a]
        );
        compare(
          `b${b}.h${h}.qkvSliceV[${t},${a}]`,
          jb.qkv.data[t * 3 * C + 2 * C + h * A + a],
          rh.v[t][a]
        );
        // concat order drawn in "Head outputs combined" must be head-major
        compare(
          `b${b}.h${h}.concat[${t},${a}]`,
          jb.vCombined.data[t * C + h * A + a],
          rh.out[t][a]
        );
      }
      let rowSum = 0;
      for (let j = 0; j < T; j++) {
        compare(`b${b}.h${h}.scores[${t},${j}]`, jh.scores.data[t * T + j], rh.scores[t][j]);
        compare(`b${b}.h${h}.pattern[${t},${j}]`, jh.sm.data[t * T + j], rh.pattern[t][j]);
        const p = jh.sm.data[t * T + j];
        if (isFinite(p)) {
          rowSum += p;
          if (j > t) {
            failures++;
            console.error(`CAUSALITY VIOLATION: b${b}.h${h}.pattern[${t},${j}] is unmasked`);
          }
        }
      }
      if (Math.abs(rowSum - 1) > 1e-5) {
        failures++;
        console.error(`SOFTMAX ROW b${b}.h${h}[${t}] sums to ${rowSum}`);
      }
    }
  }
  for (let t = 0; t < T; t++) {
    for (let c = 0; c < C; c++) {
      compare(`b${b}.proj[${t},${c}]`, jb.attnProj.data[t * C + c], rb.proj[t][c]);
      compare(`b${b}.resid[${t},${c}]`, jb.attnResid.data[t * C + c], rb.resid[t][c]);
    }
  }
}

const nBlocks = ref.blocks.length;
console.log(`checked ${nBlocks} blocks x ${H} heads x T=${T}`);
console.log(`worst abs error: ${worst.err.toExponential(3)} at ${worst.where}`);
if (failures > 0) {
  console.error(`FAILED: ${failures} mismatches`);
  process.exit(1);
}
console.log("OK: attention internals match PyTorch exactly (within float32)");
